### PR TITLE
Inline Help: add `location` property to records tracks event

### PR DESCRIPTION
**Pull request by retrofox**
#### Changes proposed in this Pull Request

This PR improves recording tracks event from the Inline Help Popover component (currently working on prod) and also records the `search` event when a search occurs in the `<Support />` component.

~<Support /> isn't implemented yet. It's currently [handled on this PR](https://github.com/Automattic/wp-calypso/pull/42625). For this reason, this branch was been branched off from there.~

### Segmentation

The primary goal here is to be able to compare where the events came from, segmenting them by the new `location` property, which could contain two different values so far: `inline-help-popover` and `customer-home`.
These values make reference to the location where the search was done by the user.

### Scope

The suggested scope is _"Record track event when the user performs a search query in the customer home, and add the `location` property to all events of the support assistant"_. It will allow us to compare stats between the customer home page and the rest (inline help popover).

### Next Steps

Once this approach merges, the next steps are:

* Keep recording more tracks event with the proper segmentation
* Re-evaluate if some of events could be combined into a single event like `calypso_inlinehelp_resource_open|view|click|hide` with property `resource=contact|morehelp|video|tour|forums`

### Testing instructions

1) Apply D43941-code to your sandbox (for testing only).
2) Set up the `debug` property of the local storage to see the recording events.
```
localStorage.setItem( 'debug', 'calypso:analytics*' );
```

3) Check that events triggered and recorded by the <InlineHelpPopover /> have the `location` property with the `inline-help-popover` value. 

* Go to any page different to customer home, for instance, stats
http://calypso.localhost:3000/stats/day/<your-testing-site>
* Open the Inline Help Popover by clicking on the FAB icon
* Tyle a query in the _🔎 Search for help..._ input
* Confirm that the events have the `location` property in the dev console.

_Trigger to record the event_
Screen Shot 2020-05-28 at 8 39 49 AM

_Recorded event_
Screen Shot 2020-05-28 at 8 40 16 AM

4) Check the events recorded by the `<Support />` component, available in the customer home page:

* http://calypso.localhost:3000/home/<your-testing-site>
* Tyle a query in the _🔎 Search for help..._ input
* Confirm that the events have the `location` property in the dev console.

Screen Shot 2020-05-28 at 9 01 13 AM

Screen Shot 2020-05-28 at 9 01 20 AM

5) Also, you can confirm that it doesn't record events for empty query strings.

Fixes https://github.com/Automattic/wp-calypso/issues/42105

### CODE REVIEW

**Comment by gwwar** [comment]
@retrofox see PCYsg-q5t-p2 for customer home analytics events. I believe we'll want to group this under calypso_cutomer_home as a prefix for consistency

**Comment by retrofox** [comment]
> @retrofox see PCYsg-q5t-p2 for customer home analytics events. I believe we'll want to group this under calypso_cutomer_home as a prefix for consistency

Thanks @gwwar for the ping. I've left a comment in _feature_ PR about this. Since we've already started to discuss it there, let's keep it there too until we get a final decision.

**Comment by retrofox** [comment]
> > @retrofox see PCYsg-q5t-p2 for customer home analytics events. I believe we'll want to group this under calypso_cutomer_home as a prefix for consistency
> 
> Thanks @gwwar for the ping. I've left a comment in _feature_ PR about this. Since we've already started to discuss it there, let's keep it there too until we get a final decision.

[link](https://github.com/Automattic/wp-calypso/pull/42625#discussion_r430746329)

**Comment by obenland** [comment]
@retrofox When you get a chance, could you summarize the latest status here and next steps?

**Comment by retrofox** [comment]
> @retrofox When you get a chance, could you summarize the latest status here and next steps?

The PR describes the latest status of it. Let me add the next steps.

**Comment by retrofox** [comment]
@gwwar @martinremy 👋 

Hi there!
In order to get a consensus, we'd like to know your opinion about this approach. Basically the thing to debate is whether we should keep the same event name for both components and segment via a new prop, or set a different name for each location.

File: client/my-sites/customer-home/cards/features/help-search/search-card.jsx [hunk 1]
```diff
@@ -103,9 +87,23 @@ const mapStateToProps = ( state, ownProps ) => ( {
 	selectedResultIndex: getSelectedResultIndex( state ),
 	selectedResult: getInlineHelpCurrentlySelectedResult( state ),
 } );
+
+const requestInlineSearchResultsAndTrack = ( searchQuery ) =>
+	( ! searchQuery || ! ( searchQuery.trim() ).length )
+		? requestInlineHelpSearchResults()
+		: withAnalytics(
+			composeAnalytics (
+				recordTracksEvent( 'calypso_inlinehelp_search', {
```
Comment (gwwar) [inline client/my-sites/customer-home/cards/features/help-search/search-card.jsx:96]: @retrofox I don't have super strong feelings on this, but I do think it'd probably make sense for this one to be `calypso_customer_home_support_card_search` or similar and drop the location prop.

Were folks not interested in a bump stat?

Comment (retrofox) [reply client/my-sites/customer-home/cards/features/help-search/search-card.jsx:96]: Thanks, Kerry for your answer.

> I do think it'd probably make sense for this one to be calypso_customer_home_support_card_search or similar for discoverability and drop the location prop.

The goal for our purpose is what's the best way to compare stats between these two components. Using an event property allows us to segment the stats which, I think, makes things easier. Compares different events maybe could get trickier.



> We could keep them grouped if we were interested in comparing searches from inline vs customer home.

Could you elaborate on this idea, please?

> Also, were folks not interested in a bump stat?

AFAIK, we haven't added a bump stats in the current Inline Help Popover component, so for this reason I hadn't added to this one either. Asking just in case...

**Comment by jancavan** [comment]
@retrofox Will we be able to segment the inline help/FAB per location too?

**Comment by retrofox** [comment]
> @retrofox Will we be able to segment the inline help/FAB per location too?

Yes, we will be able, Jan. In this case, the location value is `inline-help-popover'.

File: client/my-sites/customer-home/cards/features/help-search/search-card.jsx (hunk 1 above)
Comment (gwwar) [reply client/my-sites/customer-home/cards/features/help-search/search-card.jsx:96]: > Could you elaborate on this idea, please?

😆 you just screencapped that yes, you would like to compare searches. If that's the case feel free to keep the grouping but you may want to update documentation for folks.

**Comment by retrofox** [comment]
> @retrofox Will we be able to segment the inline help/FAB per location too?

Sorry, I think I misunderstood the question. Thinking well we wondering now if we should set the location for all app sections, such as `pages`, `posts`, `earn`, `marketing`, `media`, whatever, or simply leave it as it and iterate later.

cc @davemart-in

**Comment by jancavan** [comment]
> Will we be able to segment the inline help/FAB per location too?

So I thought about this question because I was wondering if simply comparing `inline-help-popover` vs `customer-home` would make for a fair comparison. Considering the fact that the `inline-help-popover` is located in multiple places, I'm assuming that that'll naturally inflate the numbers since we're looking at an aggregate. My assumption could be wrong though.

> Thinking well we wondering now if we should set the location for all app sections, such as pages, posts, earn, marketing, media, whatever

Thanks @retrofox, I think this could be valuable. My comment above isn't a blocker though and can be handled separately, but wanted to just throw it out there.

cc @davemart-in

**Comment by obenland** [comment]
@retrofox Is there anything I can do to help move this forward?

**Comment by retrofox** [comment]
> @retrofox Is there anything I can do to help move this forward?

Test and confirm that it works as expected. I'd like to know the @getdave opinion too since he has been working on this, confirming that the `<Support />` component is not shipped to prod yet, for instance.

**Comment by getdave** [comment]
Still testing this @retrofox

File: client/my-sites/customer-home/cards/features/help-search/search-card.jsx [hunk 2]
```diff
@@ -103,9 +87,23 @@ const mapStateToProps = ( state, ownProps ) => ( {
 	selectedResultIndex: getSelectedResultIndex( state ),
 	selectedResult: getInlineHelpCurrentlySelectedResult( state ),
 } );
+
+const requestInlineSearchResultsAndTrack = ( searchQuery ) =>
```
Comment (getdave) [inline client/my-sites/customer-home/cards/features/help-search/search-card.jsx:91]: Combining everything here might be difficult to test in the future. Would it be better to have separate concerns here for "requesting results" and "tracking the query". 

Indeed the actual result of calling `requestInlineSearchResultsAndTrack` doesn't always result in events being tracked.

**Comment by retrofox** [comment]
Thanks @marekhrabe and @getdave fr the review

> I did notice however that we're not tracking the opening of the Support docs dialog/modal view. 

I defined the **_scope_** of this PR just tracking the search (there is a section in the PR description). It will be enough to compare both components/sections. I'm trying to limit it in order to make it easier to ship. I thought we could add the other steps in the near future. For this, I've written the _**Next Steps**_ section too.

You'll want to look at adding this here:
> 
> https://github.com/Automattic/wp-calypso/blob/bfb18cbbef84d8dfb8a5237bfe6a64b7c4b83598/client/my-sites/customer-home/cards/features/help-search/index.jsx#L45-L63
> 
> This is the equivalent of this from the FAB implementation:
> 
> https://github.com/Automattic/wp-calypso/blob/bfb18cbbef84d8dfb8a5237bfe6a64b7c4b83598/client/blocks/inline-help/popover.jsx#L109-L115
> 
> You may also need to add tracking to the event of clicking the "More help" link in the footer:
> 
> https://github.com/Automattic/wp-calypso/blob/bfb18cbbef84d8dfb8a5237bfe6a64b7c4b83598/client/my-sites/customer-home/cards/features/help-search/index.jsx#L77-L85

File: client/my-sites/customer-home/cards/features/help-search/search-card.jsx (hunk 2 above)
Comment (retrofox) [reply client/my-sites/customer-home/cards/features/help-search/search-card.jsx:91]: > Indeed the actual result of calling requestInlineSearchResultsAndTrack doesn't always result in events being tracked.

It doesn't track only when the query is not defined or it is an empty string, which is intentional. Is it correct to you?
